Lib/concurrent/futures/_base.py

```diff
@@ -594,10 +599,21 @@ def map(self, fn, *iterables, timeout=None, chunksize=1):
                 before the given timeout.
             Exception: If fn(*args) raises for any values.
         """
+        if buffersize is not None and buffersize < 1:
+            raise ValueError("buffersize must be None or >= 1.")
+
         if timeout is not None:
             end_time = timeout + time.monotonic()
 
-        fs = [self.submit(fn, *args) for args in zip(*iterables)]
+        args_iter = iter(zip(*iterables))
+        if buffersize:
+            fs = collections.deque(
+                self.submit(fn, *args) for args in islice(args_iter, buffersize)
+            )
+        else:
+            fs = [self.submit(fn, *args) for args in args_iter]
+
+        executor_weakref = weakref.ref(self)
```

Comment:
Do you really need a weak reference?

Reply:
It definitely needs a comment to explain why, will do that asap.

Reply:
The comment is ok. But it's only if you are actually keeping a strong reference on the threads and try to something with them right? So I'm wondering whether we should be fool-proofed against such case.

Reply:
I agree that it feels like too much precaution and I don't have neither a proper real world scenario where this is a requirement. But because `test_del_shutdown` is an existing test added 14 years ago we have to consider that this guarantee is required by some code bases, right? We don't want a potential usage of buffersize in those code bases to break it in a really hard to debug way (no error, program just stuck). And the use of the weakref has no obvious downsides, or do you find one?

Reply:
> But because test_del_shutdown is an existing test added 14 years ago we have to consider that this guarantee is required by some code bases, right?

This is sufficient for me. So let's keep this weakref. Thanks

Reply:
great 🤝